src/libcore/future/pending.rs

```diff
@@ -0,0 +1,59 @@
+use crate::future::Future;
+use crate::marker;
+use crate::pin::Pin;
+use crate::task::{Context, Poll};
+
+/// Creates a future which never resolves, representing a computation that never
+/// finishes.
+///
+/// This `struct` is created by the [`pending`] function. See its
+/// documentation for more.
+///
+/// [`pending`]: fn.pending.html
+#[unstable(feature = "future_readiness_fns", issue = "70921")]
+#[derive(Debug)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct Pending<T> {
+    // We only ever produce T. By marking we don't store it we can bypass
+    // dropck and be covariant.
+    _data: marker::PhantomData<dyn Fn() -> T>,
```

Comment:
Isn't `dyn Fn() -> T` invariant over `T` because it's an associated type constraint?

We don't do this anywhere else AFAICT, there's a `PhantomData<fn()  -> T>` in `core::mem::Discriminant` but no `PhantomData<Fn` or `PhantomData<dyn Fn` in the codebase.

Reply:
Btw variance is pretty easy to test for if you keep it simple:
* [`PhantomData<dyn Fn() -> T>` is invariant](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2e9b292f39f6ef45d02eae461c7b3c82)
* [`PhantomData<fn() -> T>` is covariant](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=32a5d5cef98dc90a5842e0b3e6476005)

Reply:
`futures::future::Pending<T>` uses `PhantomData<T>` so it's covariant (but also interacts with dropck).

`async-std` uses `pub async fn pending<T>() -> T` which means it's necessarily invariant (although the private `struct Pending<T>`, that it relies on, looks identical to the `futures` one).

Given that a value of `T` never exists at any point and the parameter only exists for trait dispatch (I don't know what "We only ever produce T." means in the comment), I'd argue that `T` could be bivariant here, but that's not an option nowadays (cc @nikomatsakis this may be an interesting edge case).

Reply:
This comment was added based on @matklad's review in https://github.com/rust-lang/rust/pull/70834#discussion_r403975529. I don't think I have enough understanding of variance to participate in this discussion. Is there a change we should make here to enable this PR to move forward?

Reply:
I think `fn() -> T` is probably what we want here, based on the comment anyway.

Reply:
@yoshuawuyts Ah, @matklad's comment is entirely correct, but note that it said `fn() -> T`, not `dyn Fn() -> T`. If I had to guess, maybe `fn` was typo'd as `Fn` and then the compiler asked for `dyn` to be added for the new, more explicit, trait object syntax.

Like @nikomatsakis said, `dyn Fn` needs to be replaced with `fn`.

<hr/>

`PhantomData<fn() -> T>` and `PhantomData<T>` have the same ("normal" aka "co-") variance, it's just that the former is more relaxed in terms of reasoning about drops (whereas the latter would be *needed* if you had, say,  an owning raw pointer to a heap-allocated `T` value, like `Box<T>` or even `Rc<T>`).

<hr/>

`PhantomData<dyn Fn() -> T>`, OTOH, being *a trait object*, is necessarily (because we know nothing of the concrete type, which may have arbitrary restrictions) invariant over `T` .
That makes it similar to `PhantomData<Cell<T>>` and `PhantomData<&'a mut T>`.

If `Pending<T>` was invariant over `T`, then  `Pending<&'static ()>` would not be able to become `Pending<&'a ()>` for some `'a` smaller than `'static`.

Reply:
That said, we just saw in the context of `Discriminant<T>` that we may wish we had been invariant here in the future. Perhaps we should leave it as is. (For that matter, I suspect that `PhantomData<T>` would suffice, even if it's more conservative with respect to dropck.) Seems like best practices here are not entirely obvious.

Reply:
> I suspect that `PhantomData<T>` would suffice (...) best practices here are not entirely obvious.

Alright; moved the PR to use `PhantomData<T>` since that also happens to be the simplest impl. All of this is internal anyway and doesn't risk having accidental soundness implications; it can always be changed later. I think at the very least this should be good enough to release on nightly.